src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp

```diff
@@ -467,6 +483,7 @@ float SampleHandle::FindSampleEditingLevel
    if (env)
    {
       // Calculate sample as it would be rendered, so quantize time
+      // todo(mhodgkinson) does this need stretch ratio?
       double envValue = env->GetValue( t0, 1.0 / mClickedTrack->GetRate());
```

Comment:
Good question, and I should have asked it too about the other use of `Envelope::GetValue` in this file.

The previous commit "Project tempo changes affect WaveClip stretch ratios ..." updates the times and durations stored directly in the WaveClip immediately whenever there is a tempo change.  And presumably that will happen too when adjustment of individual clips is implemented.  And therefore when you hit-test clips by time, no additional accounting for ratio is needed.

But you did not, in that commit, to anything to adjust the envelope values.

Either you should do that, or else adjust time with ratios here.

Maybe it's easier to understand to update all times and durations, in Envelope as well as WaveClip, to complete this.

And maybe that would incidentally fix other misbehavior of envelope display that you showed me yesterday?

Reply:
> Maybe it's easier to understand to update all times and durations, in Envelope as well as WaveClip, to complete this.

Indeed this is the approach taken, in that one-liner commit, calling `mEnvelope->RescaleTimesBy(ratioChange)` when project tempo change is reported.